File: plugins/tissuumaps_wsi_qc/WSI_QC.py
"""Server side of the WSI QC TissUUmaps plugin.

This module exposes a single API endpoint ``run_pipeline`` which executes the
quality‑control pipeline on the server and writes TissUUmaps‑ready overlay CSVs
to the output directory.  The JavaScript front end triggers this endpoint and
then adds the generated overlays as layers.
"""

import logging
import os
import subprocess

from flask import abort, make_response


class Plugin:
    """TissUUmaps server plugin for WSI QC."""

    def __init__(self, app):
        self.app = app

    def run_pipeline(self, json_param):
        """Run the QC pipeline with parameters provided by the UI.

        Parameters
        ----------
        json_param: dict
            Parameters forwarded from the JavaScript side.  Expected keys are
            ``wsi`` (path to input slide), ``out`` (output directory),
            ``ref`` (optional stain reference JSON), ``tile`` (tile size), and
            ``qc_thresh``.  Only ``wsi`` and ``out`` are mandatory.

        Returns
        -------
        flask.Response
            Simple success message if the pipeline finished without errors.
        """

        if not json_param:
            logging.error("No arguments supplied to run_pipeline")
            abort(500)

        wsi = json_param.get("wsi")
        out_dir = json_param.get("out")
        ref = json_param.get("ref")
        tile = str(json_param.get("tile", 512))
        qc_thresh = str(json_param.get("qc_thresh", 0.80))

        if not wsi or not out_dir:
            logging.error("Missing required arguments: wsi=%s, out_dir=%s", wsi, out_dir)
            abort(500)

        tiles_dir = os.path.join(out_dir, "tiles")
        qc_csv = os.path.join(out_dir, "qc.csv")

        try:
            subprocess.run([
                "python",
                "scripts/tile_wsi.py",
                "--wsi",
                wsi,
                "--out",
                out_dir,
                "--tile",
                tile,
            ], check=True)

            subprocess.run([
                "python",
                "scripts/qc_score_tiles.py",
                "--tiles",
                tiles_dir,
                "--out",
                qc_csv,
            ], check=True)

            for metric, name in [
                ("qc_score", "qc"),
                ("stripe_score", "stripe"),
                ("bubble_score", "bubble"),
                ("fold_score", "fold"),
            ]:
                subprocess.run([
                    "python",
                    "scripts/export_heatmap_to_tissuumaps.py",
                    "--tile_csv",
                    qc_csv,
                    "--metric",
                    metric,
                    "--out",
                    os.path.join(out_dir, f"overlay_{name}.csv"),
                ], check=True)

        except subprocess.CalledProcessError:
            logging.exception("QC pipeline failed")
            abort(500)

        return make_response("QC pipeline completed")

File: plugins/tissuumaps_wsi_qc/WSI_QC.js
/**
 * Front-end part of the WSI QC TissUUmaps plugin.
 *
 * The panel exposes a few parameters to run the QC pipeline and loads the
 * resulting overlay CSV files as separate layers.
 */

var WSI_QC;

WSI_QC = {
  name: "WSI QC",
  parameters: {
    _section_main: {
      label: "Pipeline",
      title: "Run QC pipeline",
      type: "section",
      collapsed: false,
    },
    _wsi: {
      label: "WSI (.svs)",
      type: "text",
      default: "",
    },
    _out: {
      label: "Output dir",
      type: "text",
      default: "",
    },
    _ref: {
      label: "Ref JSON",
      type: "text",
      default: "",
    },
    _tile: {
      label: "Tile size",
      type: "number",
      default: 512,
    },
    _qc_thresh: {
      label: "QC threshold",
      type: "number",
      default: 0.8,
      attributes: { step: 0.05, min: 0, max: 1 },
    },
    _run: {
      label: "Run",
      type: "button",
    },
    _section_layers: {
      label: "Layers",
      title: "Overlay display",
      type: "section",
      collapsed: false,
    },
    _show_qc: {
      label: "QC score",
      type: "checkbox",
      default: true,
    },
    _show_stripe: {
      label: "Stripe",
      type: "checkbox",
      default: false,
    },
    _show_bubble: {
      label: "Bubble",
      type: "checkbox",
      default: false,
    },
    _show_fold: {
      label: "Fold",
      type: "checkbox",
      default: false,
    },
  },
};

WSI_QC.init = function (container) {
  interfaceUtils.alert("WSI QC plugin loaded");
};

WSI_QC.inputTrigger = function (input) {
  if (input === "_run") {
    let payload = {
      wsi: WSI_QC.get("_wsi"),
      out: WSI_QC.get("_out"),
      ref: WSI_QC.get("_ref"),
      tile: WSI_QC.get("_tile"),
      qc_thresh: WSI_QC.get("_qc_thresh"),
      show: {
        qc: WSI_QC.get("_show_qc"),
        stripe: WSI_QC.get("_show_stripe"),
        bubble: WSI_QC.get("_show_bubble"),
        fold: WSI_QC.get("_show_fold"),
      },
    };

    let success = function (resp) {
      interfaceUtils.alert(resp);
      if (payload.show.qc) {
        interfaceUtils.addLayerFromUrl({
          url: payload.out + "/overlay_qc.csv",
          name: "QC score",
          type: "csv",
        });
      }
      if (payload.show.stripe) {
        interfaceUtils.addLayerFromUrl({
          url: payload.out + "/overlay_stripe.csv",
          name: "Stripe",
          type: "csv",
        });
      }
      if (payload.show.bubble) {
        interfaceUtils.addLayerFromUrl({
          url: payload.out + "/overlay_bubble.csv",
          name: "Bubble",
          type: "csv",
        });
      }
      if (payload.show.fold) {
        interfaceUtils.addLayerFromUrl({
          url: payload.out + "/overlay_fold.csv",
          name: "Fold",
          type: "csv",
        });
      }
    };

    let error = function (err) {
      console.log("WSI_QC error", err);
      interfaceUtils.alert("WSI QC failed");
    };

    WSI_QC.api("run_pipeline", payload, success, error);
  }
};
